plugin/trino-druid/src/test/java/io/trino/plugin/druid/TestDruidTypeMapping.java

```diff
@@ -170,50 +180,76 @@ public void testVarchar()
-                {"1970-01-01 00:00:00.123999999", "TIMESTAMP '1970-01-01 00:00:00.123'"},
+                .add(new TimestampCase("1970-01-01 00:00:01", "TIMESTAMP '1970-01-01 00:00:01.000'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:02.1", "TIMESTAMP '1970-01-01 00:00:02.100'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:03.12", "TIMESTAMP '1970-01-01 00:00:03.120'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:04.123", "TIMESTAMP '1970-01-01 00:00:04.123'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:05.1239", "TIMESTAMP '1970-01-01 00:00:05.123'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:06.12399", "TIMESTAMP '1970-01-01 00:00:06.123'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:07.123999", "TIMESTAMP '1970-01-01 00:00:07.123'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:08.1239999", "TIMESTAMP '1970-01-01 00:00:08.123'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:09.12399999", "TIMESTAMP '1970-01-01 00:00:09.123'", id++))
+                .add(new TimestampCase("1970-01-01 00:00:00.123999999", "TIMESTAMP '1970-01-01 00:00:00.123'", id++))
                 // before epoch with second fraction
-                {"1969-12-31 23:59:59.1230000", "TIMESTAMP '1969-12-31 23:59:59.123'"}
-        };
+                .add(new TimestampCase("1969-12-31 23:59:59.1230000", "TIMESTAMP '1969-12-31 23:59:59.123'", id))
+                .build();
+
+        try (DruidTable testTable = new DruidTable("test_timestamp")) {
+            String dataFilePath = format("%s/%s.tsv", druidServer.getHostWorkingDirectory(), testTable.getName());
+            try (BufferedWriter writer = new BufferedWriter(new FileWriter(dataFilePath, UTF_8))) {
+                for (TimestampCase row : rows) {
+                    writer.write("%s\t%s".formatted(row.inputLiteral, row.id));
+                    writer.newLine();
+                }
+            }
+            String dataSource = testTable.getName();
+            IndexTaskBuilder builder = new IndexTaskBuilder();
+            builder.setDatasource(dataSource);
+            TimestampSpec timestampSpec = new TimestampSpec("dummy_druid_ts", "auto");
+            builder.setTimestampSpec(timestampSpec);
+            builder.addColumn("id", "long");
+            druidServer.ingestData(testTable.getName(), Optional.empty(), builder.build(), dataFilePath);
+
+            for (TimestampCase row : rows) {
+                assertThat(query("SELECT __time FROM druid.druid." + testTable.getName() + " WHERE id = " + row.id))
+                        .matches("VALUES " + row.expectedLiteral);
+                assertThat(query("SELECT id FROM druid.druid." + testTable.getName() + " WHERE __time = " + row.expectedLiteral))
+                        .matches("VALUES BIGINT '" + row.id + "'");
```

Comment:
do you think `.as()` or whatever equivalent TestNG has to add a description to assertions could be used to indicate which `row` it failed for? Or does the assertion failure already include the `row`?